src/main/java/io/cryostat/diagnostic/DiagnosticsHelper.java

```diff
@@ -0,0 +1,341 @@
+                        .map(
+                                e ->
+                                        Tag.builder()
+                                                .key(
+                                                        base64Url.encodeAsString(
+                                                                e.getKey()
+                                                                        .getBytes(
+                                                                                StandardCharsets
+                                                                                        .UTF_8)))
+                                                .value(
+                                                        base64Url.encodeAsString(
+                                                                e.getValue()
+                                                                        .getBytes(
+                                                                                StandardCharsets
+                                                                                        .UTF_8)))
+                                                .build())
+                        .toList());
+        return Tagging.builder().tagSet(tags).build();
+    }
+
+    public String downloadUrl(String jvmId, String filename) {
+        return String.format(
+                "/api/beta/diagnostics/threaddump/download/%s", encodedKey(jvmId, filename));
+    }
+
+    public String encodedKey(String jvmId, String uuid) {
+        Objects.requireNonNull(jvmId);
+        Objects.requireNonNull(uuid);
+        return base64Url.encodeAsString(
+                (threadDumpKey(jvmId, uuid)).getBytes(StandardCharsets.UTF_8));
+    }
+
+    public String threadDumpKey(String jvmId, String uuid) {
+        return (jvmId + "/" + uuid).strip();
+    }
+
+    public String threadDumpKey(Pair<String, String> pair) {
+        return threadDumpKey(pair.getKey(), pair.getValue());
+    }
+
+    public InputStream getThreadDumpStream(String jvmId, String threadDumpID) {
+        return getThreadDumpStream(encodedKey(jvmId, threadDumpID));
+    }
+
+    public InputStream getThreadDumpStream(String encodedKey) {
+        Pair<String, String> decodedKey = decodedKey(encodedKey);
+        var key = decodedKey.getValue().strip();
+
+        GetObjectRequest getRequest = GetObjectRequest.builder().bucket(bucket).key(key).build();
+
+        return storage.getObject(getRequest);
+    }
+
+    public Pair<String, String> decodedKey(String encodedKey) {
+        String key = new String(base64Url.decode(encodedKey), StandardCharsets.UTF_8);
+        String[] parts = key.split("/");
+        if (parts.length != 2) {
+            throw new IllegalArgumentException();
+        }
+        return Pair.of(parts[0], parts[1]);
+    }
+
+    public List<S3Object> listThreadDumps(long targetId) {
+        var jvmId = Target.<Target>findById(targetId).jvmId;
```

Comment:
Similar here with `findById` being nullable, and the `.jvmId` also being nullable.